### ISSUE TITLE
colcon build failed undefined reference to `DUtils::Random::SeedRandOnce()'

### ISSUE THREAD

**Issue by spectra543** [zang09/ORB_SLAM3_ROS2#15]
Hi guy i try to colcon build ORB_SLAM3_ROS2 after i run this line  "colcon build --symlink-install --packages-select orbslam3" this error happens.

![dtuil](https://github.com/zang09/ORB_SLAM3_ROS2/assets/140332516/e04f56df-14ef-4f20-96c0-24970e07543f)

i use 
ros2 foxy
opencv 4.2.0
ubuntu 20.04
anyone know how to fix thankyou!

**Comment by mjunsen123** [comment]
1. Go to ORB_SLAM3/Thirdparty/DBoW2/CMakeLists.txt
2. Change line `add_library(DBoW2 SHARED ${SRCS_DBOW2} ${SRCS_DUTILS})` to `add_library(DBoW2 STATIC ${SRCS_DBOW2} ${SRCS_DUTILS})` and,
3. add line `set_property(TARGET DBoW2 PROPERTY POSITION_INDEPENDENT_CODE ON)`
4. Go to ORB_SLAM3/CMakeLists.txt
5. Replace `${PROJECT_SOURCE_DIR}/Thirdparty/DBoW2/lib/libDBoW2.so` with `${PROJECT_SOURCE_DIR}/Thirdparty/DBoW2/lib/libDBoW2.a`
6. Rebuild

Hope this help!

**Comment by 1melek** [comment]
i have the same issue , i tried  doing your method and it still didn't work

**Comment by mjunsen123** [comment]
> i have the same issue , i tried doing your method and it still didn't work

Have you rebuild ORBSLAM3 as well? Do you have `.a` file in Thirdparty/DBoW2/lib/libDBoW2.a after rebuild?

**Comment by 1melek** [comment]
well no i will try to rebuild it now

**Comment by 1melek** [comment]
okay that worked ! but now the build failed again.. 

![Screenshot from 2024-05-14 12-20-23](https://github.com/zang09/ORB_SLAM3_ROS2/assets/144792577/ca193a90-25f6-4ffa-b98a-18b835716579)

 can you help?

**Comment by mjunsen123** [comment]
> okay that worked ! but now the build failed again..
> 
> ![Screenshot from 2024-05-14 12-20-23](https://private-user-images.githubusercontent.com/144792577/330565797-ca193a90-25f6-4ffa-b98a-18b835716579.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU3MTQ4NjEsIm5iZiI6MTcxNTcxNDU2MSwicGF0aCI6Ii8xNDQ3OTI1NzcvMzMwNTY1Nzk3LWNhMTkzYTkwLTI1ZjYtNGZmYS1iOThhLTE4YjgzNTcxNjU3OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxNFQxOTIyNDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zM2QyNzczMzg2NTUyMThlYzYyOGU3ZDUyODlkN2M5NjU5MzRjN2UzYjk4OTI3MzUyYWJjYmE2ZTkwYzhkMmVhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.yzFcEXiNu_6rbtbsOzfYEJ57AshhQRyOBWkqXDjWQa8)
> 
> can you help?

Your OpenCV version in ROS is different with the one in your machine. I assume you have build cvbridge, edit the file CMakeList.txt in CVBridge to force it to use OpenCV 4.9.

**Comment by 1melek** [comment]
no i haven't build CVBbridge .. should i?
i don't even know what and where it is .

**Comment by mjunsen123** [comment]
Okay if it is the case i suggest you downgrade your opencv in your ubuntu to 4.2. I belive that would be the most straightforward solution

**Comment by 1melek** [comment]
the thing is , i have opencv 4.2 and i have no idea where 4.9 is coming from
![11](https://github.com/zang09/ORB_SLAM3_ROS2/assets/144792577/b883dffc-a966-4ad7-8180-defab3de1b6d)

**Comment by mjunsen123** [comment]
Have you build with source before?

**Comment by 1melek** [comment]
I figured that and it build successfully but now when i run it i can't see anything in the camera and I don’t know from where it get camera input , like it always says “1 camera in the atlas” even when i deactivated the laptop camera and no usb cam

**Comment by andremaurell** [comment]
@mjunsen123  Good tutorial my friend, but, can you help-me? after the changes at CMake and generate .a file, I received that error:

![image](https://github.com/zang09/ORB_SLAM3_ROS2/assets/90858902/add11dad-1a93-4f49-aa42-e5847461d546)

**Comment by mjunsen123** [comment]
Never see this error before. Can you please confirm you have CVBridge (vision_opencv) installed?

**Comment by andremaurell** [comment]
I think yes:

![image](https://github.com/zang09/ORB_SLAM3_ROS2/assets/90858902/0a1a110a-69a6-4bed-8452-0e458306328d)

**Comment by mjunsen123** [comment]
And can you confirm that you have same version of OpenCV in both your ROS and Ubuntu?

**Comment by andremaurell** [comment]
i don't know how I can see on ROS, but on ubuntu is 4.5.4

**Comment by mjunsen123** [comment]
I am not sure but I think it may due to different version of openCV in Ros and Ubuntu. I am aware of this but I am not sure if this is your case. What I did is I build opencv and cvbridge from source, so I can specific the version used in ORBSLAM3

**Comment by andremaurell** [comment]
Yeah, I think is the same version, but I'm not sure about that. Do you have same tutorial showing how I can make this build?
